packages/api-derive/src/assets/all.ts

```diff
@@ -0,0 +1,85 @@
+// Copyright 2017-2022 @polkadot/api-derive authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import type { EventRecord, Hash } from '@polkadot/types/interfaces';
+import type { FrameSystemEventRecord } from '@polkadot/types/lookup';
+import type { Vec } from '@polkadot/types-codec';
+import type { DeriveApi } from '../types';
+import type { DeriveAsset, FetchedAssetsEntries, FetchedAssetsMetadataEntries } from './types';
+
+import { combineLatest, concat, EMPTY, map, Observable, of, switchMap, take } from 'rxjs';
+
+const ASSET_EVENTS: string[] = [
+  'Created',
+  'Destroyed',
+  'MetadataCleared',
+  'MetadataSet',
+  'OwnerChanged',
+  'TeamChanged',
+  'Issued'
+];
+
+export function extractAssetEventsHash (events: Vec<FrameSystemEventRecord>): Observable<Hash> {
+  const filtered = events.find(({ event: { method, section } }) => section === 'assets' && ASSET_EVENTS.includes(method));
+
+  return filtered ? of(events.createdAtHash as Hash) : EMPTY;
+}
+
+function concatAssetData ([maybeAssets, maybeMetadatas]: [maybeAssets: FetchedAssetsEntries, maybeMetadatas: FetchedAssetsMetadataEntries]): DeriveAsset[] {
+  const result: DeriveAsset[] = [];
+
+  maybeAssets.forEach(([id, asset], index) => {
+    if (asset.isSome) {
+      const { accounts, admin, approvals, freezer, isFrozen, isSufficient, issuer, minBalance, owner, sufficients, supply } = asset.unwrap();
+      const { decimals, deposit, name, symbol } = maybeMetadatas[index][1];
+
+      result.push({
+        accounts: accounts.toBn(),
+        admin,
+        approvals: approvals.toBn(),
+        decimals: decimals.toNumber(),
+        deposit: deposit.toBn(),
+        freezer,
+        id: id.args[0],
+        isFrozen: isFrozen.toHuman(),
+        isSufficient: isSufficient.toHuman(),
+        issuer,
+        minBalance: minBalance.toBn(),
+        name: name.toUtf8(),
+        owner,
+        sufficients: sufficients.toBn(),
+        supply: supply.toBn(),
+        symbol: symbol.toUtf8()
+      } as DeriveAsset);
+    }
+  });
+
+  return result;
+}
+
+/**
+ * @name all
+ * @returns An array containing all assets with metadata
+ */
+
+export function all (api: DeriveApi): Observable<DeriveAsset[]> {
+  const initialBlockHash = api.rpc.chain.subscribeNewHeads()
+    .pipe(take(1))
+    .pipe((val) => val.pipe(map(({ hash }) => hash)));
+
+  return concat(
+    initialBlockHash,
+    api.query.system.events()
+      .pipe(switchMap((events: Vec<EventRecord>) => extractAssetEventsHash(events)))
+  )
+
+    .pipe(
+      switchMap((blockHash: Hash) =>
+        combineLatest([
+          api.query.assets.asset.entriesAt(blockHash),
+          api.query.assets.metadata.entriesAt(blockHash)
+        ])
+      )
```

Comment:
This is a very, very heave operation. Basically you refresh everything. The events actually have the ids of those changed, so should pin-point to those and only update those.

(crowdloans is an example of this - it tries to not get everything each time something has changed, rather it only refreshes based on the events received. As the state grows, this approach would just become slower and slower...)

Reply:
Understood. Though still not sure how to solve it. Reason we use `entriesAt` is initially that `multi` did not get the updates from latest (non-finalized) block (https://github.com/polkadot-js/apps/issues/6874#issuecomment-1022038369). As there is no `entriesAt(blockHash, keys)`, I understand we need to load all, and then update state based on data from events (asset and metadata changes)?

Reply:
So if we can get a minimal reproducable sample with API code, can actually log that on Substrate. (There could obviously be an issue with the apps UI implementation itself)

Reply:
Take a look at this issue, which is also parachain-related - https://github.com/paritytech/substrate/issues/10743

Would love a sample on a single connection where -

(a) event is available
(b) query for that item is not updated

I honestly cannot replicate this on a stand-alone node.